File: projects/mcp-8004-enterprise/agents_implementation/client-setup-bridge.js
/**
 * client-setup-bridge.js  [BANK DMZ INTERNAL — submit-only]
 *
 * Watches OnboardingRegistry.PhaseCompleted events and drives the three
 * sequential client setup phases.
 *
 *   ALL_REVIEWS_DONE set  → setup_legal_entity → ClientSetupOracle.setupLegalEntity()
 *   ENTITY_SETUP_DONE set → setup_account      → ClientSetupOracle.setupAccount()
 *   ACCOUNT_SETUP_DONE set→ setup_products     → ClientSetupOracle.setupProducts()
 *
 * Each setup agent has its own NFT id; all share the same MCP server script.
 *
 * Usage:
 *   node client-setup-bridge.js \
 *     --onboarding-registry  0x... \
 *     --setup-contract       0x... \
 *     --rpc http://127.0.0.1:8545 \
 *     --privkey 0x... \
 *     --entity-agent-id  4 \
 *     --account-agent-id 5 \
 *     --product-agent-id 6
 *
 * ENV: ONBOARDING_REGISTRY_ADDRESS, SETUP_CONTRACT_ADDRESS, RPC_URL, ORACLE_PRIVATE_KEY,
 *      ENTITY_AGENT_ID, ACCOUNT_AGENT_ID, PRODUCT_AGENT_ID
 */

import { ethers } from 'ethers';
import { arg, callMcpTool } from './bridge-base.js';
import { createSigner }      from './vault-signer.js';

const LABEL = 'setup-bridge';

const RPC_URL          = arg('--rpc',                  'RPC_URL')                   ?? 'http://127.0.0.1:8545';
const PRIVATE_KEY      = arg('--privkey',              'ORACLE_PRIVATE_KEY');
const SIGNER_TYPE      = arg('--signer-type',          'SIGNER_TYPE')               ?? 'local';
const VAULT_URL        = arg('--vault-url',            'VAULT_URL');
const VAULT_ADDRESS    = arg('--vault-address',        'ORACLE_ADDRESS');
const MCP_ENDPOINT     = arg('--mcp-endpoint',         'SETUP_MCP_ENDPOINT')        ?? 'http://localhost:8014';

const ONBOARDING_REG   = arg('--onboarding-registry',  'ONBOARDING_REGISTRY_ADDRESS');
const SETUP_CONTRACT   = arg('--setup-contract',        'SETUP_CONTRACT_ADDRESS');
const ENTITY_AGENT_ID  = BigInt(arg('--entity-agent-id', 'ENTITY_AGENT_ID')  ?? '4');
const ACCOUNT_AGENT_ID = BigInt(arg('--account-agent-id','ACCOUNT_AGENT_ID') ?? '5');
const PRODUCT_AGENT_ID = BigInt(arg('--product-agent-id','PRODUCT_AGENT_ID') ?? '6');

if (!ONBOARDING_REG)  { console.error(`[${LABEL}] Missing --onboarding-registry`); process.exit(1); }
if (!SETUP_CONTRACT)  { console.error(`[${LABEL}] Missing --setup-contract`);       process.exit(1); }
if (!PRIVATE_KEY && SIGNER_TYPE === 'local') { console.error(`[${LABEL}] Missing --privkey`); process.exit(1); }

const REGISTRY_ABI = [
  'event PhaseCompleted(bytes32 indexed flowId, uint8 indexed phase, uint256 timestamp)',
  'function phaseBitmask(bytes32 flowId) view returns (uint8)',
  'function ALL_REVIEWS_DONE() view returns (uint8)',
  'function PHASE_ENTITY_SETUP_DONE() view returns (uint8)',
  'function PHASE_ACCOUNT_SETUP_DONE() view returns (uint8)',
];

const SETUP_ABI = [
  'function setupLegalEntity(bytes32 flowId, uint256 agentId, bytes32 entitySpecHash)',
  'function setupAccount(bytes32 flowId, uint256 agentId, bytes32 accountSpecHash)',
  'function setupProducts(bytes32 flowId, uint256 agentId, bytes32 productSpecHash)',
];

async function main() {
  const provider = new ethers.JsonRpcProvider(RPC_URL);
  const wallet   = createSigner({ type: SIGNER_TYPE, privateKey: PRIVATE_KEY, vaultUrl: VAULT_URL, address: VAULT_ADDRESS }, provider);

  const registry     = new ethers.Contract(ONBOARDING_REG, REGISTRY_ABI, wallet);
  const setupOracle  = new ethers.Contract(SETUP_CONTRACT,  SETUP_ABI,    wallet);

  const ALL_REVIEWS_DONE    = await registry.ALL_REVIEWS_DONE();
  const ENTITY_SETUP_DONE   = await registry.PHASE_ENTITY_SETUP_DONE();
  const ACCOUNT_SETUP_DONE  = await registry.PHASE_ACCOUNT_SETUP_DONE();

  const signerAddress = await wallet.getAddress();
  console.log(`[${LABEL}] Signer             : ${signerAddress}`);
  console.log(`[${LABEL}] OnboardingRegistry : ${ONBOARDING_REG}`);
  console.log(`[${LABEL}] ClientSetupOracle  : ${SETUP_CONTRACT}`);
  console.log(`[${LABEL}] MCP endpoint       : ${MCP_ENDPOINT}`);
  console.log(`[${LABEL}] EntityAgentId      : ${ENTITY_AGENT_ID}`);
  console.log(`[${LABEL}] AccountAgentId     : ${ACCOUNT_AGENT_ID}`);
  console.log(`[${LABEL}] ProductAgentId     : ${PRODUCT_AGENT_ID}`);
  console.log(`[${LABEL}] Listening for PhaseCompleted events…\n`);

  registry.on('PhaseCompleted', async (...args) => {
    const event = args[args.length - 1];
    const { flowId, phase } = event.args;
    const mask = await registry.phaseBitmask(flowId);

    console.log(`\n[${LABEL}] ← PhaseCompleted  flowId=${flowId}  phase=0x${phase.toString(16)}  mask=0x${mask.toString(16)}`);

    // ALL_REVIEWS_DONE → trigger legal entity setup
    if ((mask & ALL_REVIEWS_DONE) === ALL_REVIEWS_DONE && (mask & ENTITY_SETUP_DONE) === 0n) {
      try {
        console.log(`[${LABEL}]   → setup_legal_entity`);
        const result = await callMcpTool(MCP_ENDPOINT, 'setup_legal_entity', { flow_id: flowId }, flowId);
        const tx = await setupOracle.setupLegalEntity(flowId, ENTITY_AGENT_ID, result.entity_spec_hash);
        await tx.wait();
        console.log(`[${LABEL}]   ✓ setupLegalEntity  tx=${tx.hash}`);
      } catch (err) { console.error(`[${LABEL}]   ✗ ${err.message}`); }
      return;
    }

    // ENTITY_SETUP_DONE → trigger account setup
    if ((mask & ENTITY_SETUP_DONE) === ENTITY_SETUP_DONE && (mask & ACCOUNT_SETUP_DONE) === 0n) {
      try {
        console.log(`[${LABEL}]   → setup_account`);
        const result = await callMcpTool(MCP_ENDPOINT, 'setup_account', { flow_id: flowId }, flowId);
        const tx = await setupOracle.setupAccount(flowId, ACCOUNT_AGENT_ID, result.account_spec_hash);
        await tx.wait();
        console.log(`[${LABEL}]   ✓ setupAccount  tx=${tx.hash}`);
      } catch (err) { console.error(`[${LABEL}]   ✗ ${err.message}`); }
      return;
    }

    // ACCOUNT_SETUP_DONE → trigger product setup
    const PRODUCT_SETUP_DONE = 0x20n;
    if ((mask & ACCOUNT_SETUP_DONE) === ACCOUNT_SETUP_DONE && (mask & PRODUCT_SETUP_DONE) === 0n) {
      try {
        console.log(`[${LABEL}]   → setup_products`);
        const result = await callMcpTool(MCP_ENDPOINT, 'setup_products', { flow_id: flowId }, flowId);
        const tx = await setupOracle.setupProducts(flowId, PRODUCT_AGENT_ID, result.product_spec_hash);
        await tx.wait();
        console.log(`[${LABEL}]   ✓ setupProducts  tx=${tx.hash}`);
      } catch (err) { console.error(`[${LABEL}]   ✗ ${err.message}`); }
    }
  });
}

main().catch(err => { console.error(err); process.exit(1); });

File: projects/mcp-8004-enterprise/agents_implementation_py/bridges/client_setup_bridge.py
"""
client_setup_bridge.py — Port of client-setup-bridge.js using web3.py AsyncWeb3.

Watches OnboardingRegistry.PhaseCompleted events and drives the three sequential
client setup phases.

  ALL_REVIEWS_DONE set   → setup_legal_entity → ClientSetupOracle.setupLegalEntity()
  ENTITY_SETUP_DONE set  → setup_account      → ClientSetupOracle.setupAccount()
  ACCOUNT_SETUP_DONE set → setup_products     → ClientSetupOracle.setupProducts()

Usage:
    python -m bridges.client_setup_bridge \
        --onboarding-registry 0x... \
        --setup-contract 0x... \
        --rpc http://... --privkey 0x... \
        --entity-agent-id 4 --account-agent-id 5 --product-agent-id 6
"""

from __future__ import annotations

import asyncio
import logging
import os

from web3 import Web3

from shared.abis import ONBOARDING_REGISTRY_ABI, SETUP_ORACLE_ABI
from shared.bridge_base import arg, call_mcp_tool
from shared.vault_signer import create_signer
from shared.bridge_base import build_web3

logger = logging.getLogger(__name__)

LABEL        = "setup-bridge"
MCP_ENDPOINT = arg("--mcp-endpoint", "SETUP_MCP_ENDPOINT", "http://localhost:8014")

RPC_URL        = arg("--rpc",             "RPC_URL",             "http://127.0.0.1:8545")
PRIVATE_KEY    = arg("--privkey",         "ORACLE_PRIVATE_KEY")
SIGNER_TYPE    = arg("--signer-type",     "SIGNER_TYPE",         "local")
VAULT_URL      = arg("--vault-url",       "VAULT_URL")
VAULT_ADDRESS  = arg("--vault-address",   "ORACLE_ADDRESS")

ONBOARDING_REG   = arg("--onboarding-registry",  "ONBOARDING_REGISTRY_ADDRESS")
SETUP_CONTRACT   = arg("--setup-contract",        "SETUP_CONTRACT_ADDRESS")
ENTITY_AGENT_ID  = int(arg("--entity-agent-id",  "ENTITY_AGENT_ID",  "4"))
ACCOUNT_AGENT_ID = int(arg("--account-agent-id", "ACCOUNT_AGENT_ID", "5"))
PRODUCT_AGENT_ID = int(arg("--product-agent-id", "PRODUCT_AGENT_ID", "6"))


def _hex(b) -> str:
    return b.hex() if isinstance(b, bytes) else str(b).removeprefix("0x")


def _bytes(h: str) -> bytes:
    return bytes.fromhex(h.removeprefix("0x"))


async def main() -> None:
    import sys
    if not ONBOARDING_REG:
        logger.error("[%s] Missing --onboarding-registry", LABEL); sys.exit(1)
    if not SETUP_CONTRACT:
        logger.error("[%s] Missing --setup-contract", LABEL); sys.exit(1)
    if SIGNER_TYPE == "local" and not PRIVATE_KEY:
        logger.error("[%s] Missing --privkey", LABEL); sys.exit(1)

    signer = create_signer(SIGNER_TYPE, PRIVATE_KEY, VAULT_URL, VAULT_ADDRESS)
    w3, _  = build_web3(RPC_URL, signer)

    registry    = w3.eth.contract(address=ONBOARDING_REG, abi=ONBOARDING_REGISTRY_ABI)
    setup_oracle= w3.eth.contract(address=SETUP_CONTRACT,  abi=SETUP_ORACLE_ABI)

    all_done     = await registry.functions.ALL_REVIEWS_DONE().call()
    entity_done  = await registry.functions.PHASE_ENTITY_SETUP_DONE().call()
    account_done = await registry.functions.PHASE_ACCOUNT_SETUP_DONE().call()
    product_done = 0x20

    logger.info("[%s] OnboardingRegistry : %s", LABEL, ONBOARDING_REG)
    logger.info("[%s] ClientSetupOracle  : %s", LABEL, SETUP_CONTRACT)
    logger.info("[%s] MCP endpoint       : %s", LABEL, MCP_ENDPOINT)
    logger.info("[%s] EntityAgentId      : %d", LABEL, ENTITY_AGENT_ID)
    logger.info("[%s] AccountAgentId     : %d", LABEL, ACCOUNT_AGENT_ID)
    logger.info("[%s] ProductAgentId     : %d", LABEL, PRODUCT_AGENT_ID)
    logger.info("[%s] Listening for PhaseCompleted events…", LABEL)

    phase_filter = await registry.events.PhaseCompleted.create_filter(from_block="latest")

    while True:
        for event in await phase_filter.get_new_entries():
            asyncio.create_task(_handle_phase_completed(
                registry, setup_oracle, event, all_done, entity_done, account_done, product_done,
            ))
        await asyncio.sleep(2)


async def _handle_phase_completed(registry, setup_oracle, event, all_done, entity_done, account_done, product_done) -> None:
    args    = event["args"]
    flow_id = args["flowId"]
    phase   = args["phase"]
    mask    = await registry.functions.phaseBitmask(flow_id).call()

    flow_hex = _hex(flow_id)
    logger.info(
        "\n[%s] ← PhaseCompleted  flowId=%s  phase=0x%02x  mask=0x%02x",
        LABEL, flow_hex, phase, mask,
    )

    # ALL_REVIEWS_DONE → legal entity setup
    if (mask & all_done) == all_done and (mask & entity_done) == 0:
        try:
            logger.info("[%s]   → setup_legal_entity", LABEL)
            result = await call_mcp_tool(MCP_ENDPOINT, "setup_legal_entity", {"flow_id": flow_hex}, flow_hex)
            tx = await setup_oracle.functions.setupLegalEntity(
                flow_id, ENTITY_AGENT_ID, _bytes(result["entity_spec_hash"])
            ).transact()
            logger.info("[%s]   ✓ setupLegalEntity  tx=%s", LABEL, tx.hex())
        except Exception as exc:
            logger.error("[%s]   ✗ %s", LABEL, exc)
        return

    # ENTITY_SETUP_DONE → account setup
    if (mask & entity_done) == entity_done and (mask & account_done) == 0:
        try:
            logger.info("[%s]   → setup_account", LABEL)
            result = await call_mcp_tool(MCP_ENDPOINT, "setup_account", {"flow_id": flow_hex}, flow_hex)
            tx = await setup_oracle.functions.setupAccount(
                flow_id, ACCOUNT_AGENT_ID, _bytes(result["account_spec_hash"])
            ).transact()
            logger.info("[%s]   ✓ setupAccount  tx=%s", LABEL, tx.hex())
        except Exception as exc:
            logger.error("[%s]   ✗ %s", LABEL, exc)
        return

    # ACCOUNT_SETUP_DONE → product setup
    if (mask & account_done) == account_done and (mask & product_done) == 0:
        try:
            logger.info("[%s]   → setup_products", LABEL)
            result = await call_mcp_tool(MCP_ENDPOINT, "setup_products", {"flow_id": flow_hex}, flow_hex)
            tx = await setup_oracle.functions.setupProducts(
                flow_id, PRODUCT_AGENT_ID, _bytes(result["product_spec_hash"])
            ).transact()
            logger.info("[%s]   ✓ setupProducts  tx=%s", LABEL, tx.hex())
        except Exception as exc:
            logger.error("[%s]   ✗ %s", LABEL, exc)


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(main())
